### DISCUSSION TITLE
Mutability vs Immutability of table defintions

### DISCUSSION THREAD

**Discussion by mathemancer** [mathesar-foundation/mathesar#3] [Maintainer discussions]
(Ported over from `#mathesar:matrix.org`)

Thinking about mutability of table definitions:  Should table definitions be
truly mutable?  For this, by "definition", I mean the specification of the
columns and constraints of the table, not the data in the rows.

### Motivating example
Should a function "move column from table _A_ to table _B_" actually mutate the
tables _A_ and _B_, or create _A'_, _B'_ with the column in the new place?

There are kind of two levels here, by the way, the user-visible level, and the
actual data layer.  It's possible to have the user think they're
mutating tables without actually mutating them

### Mutability Pros
- Data consistency will be simpler (i.e., we won't have to keep _A, A'_ in sync
  with each other)
- Less data to store, since repetition is reduced
- Easier mental model for users (I think; might be worth asking)

### Mutability Cons
- Dependencies will be broken all the time (or we'll have lots of work to
  unbreak them).
- Input that was formatted to go into table _A_ won't go into table _A'_.
- A job that reads table _A_ won't be able to read table _A'_ without changing
  the reader function.
- We'd need some elaborate system of auto-generated triggers (probably
  generated upon any mutation) to help users manage these dependencies.

### Immutability Pros:
- No dependency breakage.
- Somewhat simpler implementation, since even internal dependencies (foreign
  keys, for example) aren't broken, we don't have to shuffle identifiers
  around, etc.

### Immutability Cons:
- Data explosion
- Keeping things consistent would also take a lot of work, probably using
  triggers as well
  - For example, any write to table _A_ would need to trigger a write to _A'_
    and the moved column in _B'_. It may not be possible to figure out what to
    do about the rest of the row in _B'_ in that situation. A write to _B'_
    would be even worse (what to do about _A_?)

## Functions (so far) for which this is relevant:
- Extracting columns from a table to make a new table connected by a Foreign
  Key
- Creating a table from a join of two tables
- Moving a column from table _A_ to table _B_

I'm currently thinking that the question might have different answers for
different functions, but that inconsistency bothers me quite a bit.  Even
worse, it seems like there are only some of the users' tables for which
immutability of table definition would be useful (for example, the initial
import format, or any tables they're reading from / importing to in some
automatic way).

## Usability supreme
My dream for usability would be that a user could mark a table as "depended
upon", we'd be able to keep that table definition immutable, and the data
consistency would "just work" from the user perspective with minimal decisions
on their part.  This would be very challenging to implement.

## Previous Experience...
...indicates having some data model stability on the interface layer is useful.

One thing I keep thinking about is that the main blocker for updating the data
model of the `images` table in CC Catalog wasn't that we didn't understand what
a better model would look like, or that the actual migration would be difficult
for the table data. It was that migrating the dependencies (i.e., the various
jobs that read/write from/to that table) would have been a major cost. Solving
that part in a sensible, easy-to-use way would gain a whole class of users with
that exact problem (I've experienced it at just about every job). The data
model turns into (or gets used as) an API that needs to appear stable at some
point.

## Use case considerations

To me, the split is less "advanced" vs. "non-advanced", and more "hand input"
vs. "import from some source", or on the other side,  "ad-hoc analysis" vs.
"updating charts as more data comes in".

I do think that having the user perspective be mutable might be easier for them
conceptually (I.e., the tables would mutate for the user when appropriate)

## Current implementation (or prototype)
Implementation-wise, I've been treating tables as immutable so far, since
it's much easier to make sure everything is correct (Even the moving column
from _A_ to _B_ involves creating new _A'_ and _B'_ tables under the hood in
the current implementation, then pointing the names _A_ and _B_ to the new
tables)

## Silly idea
Another option would be to consider tables Immutable in both dimensions (i.e.,
adding a row to a table generates a new table). I think this is
intractable for obvious reasons (would we create a new table for each insert?
It seems like an RDBMS is the wrong tool to implement something that has the
actual data unchanging).  However, this would solve the consistency problems,
and is how some other systems for data wrangling work, so bears mentioning.

**Comment by metasj** [comment]
It seems to me a query/lens L into data in various tables T_i looks like a table itself, but there is no way to "write to" it.  It is rendered on read (or shows you the latest cache and its timestamp and lets you rerender if you want)

You might additionally choose to make a snapshot S or the result of viewing a lens at some timestamp; that is definitely a new immutable record, S(L,t).  You might use that snapshot to update a table T' that can itself be edited -- in which case only the version of T' at time t would be synced with a version of S.  

Do you have any cases in mind that aren't one of these three -- queries (L), permalinkable snapshots of queries (S), or tables with a series of revisions over time (T)?  I would vote for keeping every revision of T accessible, so those are essentially immutable historical tables -- even if stored in a compact way.

**Reply by mathemancer** [reply]
I think one source of confusion around this question is that we don't intend the data in the tables to be immutable.  That is, one could add or modify records.  We do intend at some point to record every entry into a table in an audit table, i.e., you should be able to reconstruct the version of T at some time t given the data in the DB, but it may not be stored as a table, exactly.  It's the _type_ of the table where immutability is in question, i.e., the list of its columns and the data types they contain.

I think the difficult thing is something like S or L, but writable.  That is, a user might start with one huge table imported from a spreadsheet.  This isn't a great data model for many applications, so they might take a pair of columns that contain repetitive blocks of data and extract them to create a new table, and create a remainder table with the remaining columns.  These tables would be linked by a foreign key relationship.  The use of the new data model would be that the user could write into it, and the data model would avoid repetitive entry, and (with the proper fkey setup) avoid mistakes.  

The question (in this example) is whether we should consider the remainder table to be the same table as the original table.  I.e., have we "removed columns" from a table, or created a new table?  The pros/cons listed above are trying to address this problem.  In particular, I expect the ultimate in flexibility would be for the user to be able to write into the original table, the extracted table, or the remainder table, and everything would be in sync and "just work".  This will get pretty complex.

Correct me if I'm wrong, but I think this falls outside of the cases you've listed.

**Comment by kgodey** [comment]
I've got a couple of thoughts here:
- We should not make a blanket design decision on mutability vs. immutability, the architecture should serve the needs of the product.
- From a user's perspective, if they change the structure of a table, they should see the table mutate.
- From the backend perspective, it seems easiest and safest to treat table structure as immutable (create new tables and rename them to match the old names).
  -  I think we should then delete the old tables for simplicity. We are planning on implementing "undo/redo" functionality and "sync from source" functionality eventually, and we should figure out how best to architect those when we come to them. It may be that keeping old tables around is a good way to implement that, but there are also other ways. It seems like premature optimization to keep the tables around.

**Comment by jbruni** [comment]
I'm not sure if what I am thinking is relevant or irrelevant, feasible or unfeasible, and not even if I truly acknowledged the point of the ongoing discussion here. I will try to share my thoughts, anyway.

What is the "**table definition**" model? We can think about it as an ordered set/list of columns, each one with properties like name, type, length, nullable, default value, etc. There may have foreign relationships, indexes, etc.

May I suggest another model. Another way to reason about what a **table definition** is, or looks like. The model described in the paragraph above would be a _snapshot_ of the model which will be described below: the "table definition **as a history**".

Rather than considering the table definition a "snapshot" of a situation which may change within time (either by "mutating" it or keeping it "immutable" and generating a new, modified, copy), we can consider the table definition as a **sequence of commands**:

1. Add "ID", integer, auto-increment, non-nullable
2. Add "Project Name", string, length 255, nullable
3. Rename "Project Name" as "Project Title"

And so on.

If we call each step a "revision", then our "table definition" is built by "revisions" (sequential commands, building a representation which can be undone).

Under this model, "deleting the old revisions" is not an option. The "undo/redo" comes out of the box. A change in the structure corresponds to appending to the revisions history. The past is immutable. A new revision creates a new table definition.

But the underlying data does not need to change. We can separate the "table definition" from the data-storage-table itself (more on this later).

There is no need to "create new tables and rename them". You just append a new command to the revision history.

And, ultimately, it allows "_for the user to be able to write into the original table, the extracted table, or the remainder table, and everything would be in sync and 'just work'_" (as mentioned as the "ultimate in flexibility", [here](https://github.com/centerofci/mathesar/discussions/3#discussioncomment-502497)). More on this later. By now, let's keep seeing the "table definitions" by themselves:

Just like in "git branches", one can start a new table definition with a command/revision which refers to another table's definition revision.

For example - "extracted table" definition:

1. Pick revision 3 from "original table"
2. Use "Project Title"

The "remainder table" could be, in fact, a new revision of the "original table":

4. Remove "Project Title"

Or another "fork" of it:

1. Pick revision 3 from "original table"
2. Use all
3. Remove "Project Title"

<hr>

There is more to it. Sure.

Under this model...

- How do you write/read data?
- How/where data is stored?

Each "revision" auto-generates its "API", mapping to "state" (or the _real data_ storage).

So, each "revision" has "getters" and "setters" which can act upon the underlying data.

Consider this table definition:

1. Add "ID", integer, auto-increment, non-nullable

We have now a "getID" and a "setID".

Next revision:

2. Add "Project Name", string, length 255, nullable

Now we have "getID", "setID", "getProjectName", "setProjectName"

3. Rename "Project Name" as "Project Title"

Now we have "getID", "setID", "getProjectTitle", "setProjectTitle". But, since this belongs to a history, "getProjectTitle" and "setProjectTitle" will map to "getProjectName" and "setProjectName".

The visualization and naming structure can have layers of getters and setters, including sorts, filters, aggregates - which, just as getters, obtain their data without actual changes to the "state" (the real data storage).

The read/write is an "API layer" as interface to interact with the underlying schema structure, and the data behind it.

The revisions can be seen as rollable-back "migrations".

Thus, you can use the read/write API of a given revision (be it "original table", "extracted table", "remainder table"), they will all map to the same "root", since the table definitions are a "history", a sequence of schema creation commands.

You gain a schema which you can move forward/backward.

This was the simple part. We can make table definitions and their revisions jump, twirl, squeeze, whirl, turn upside down, spiral. What about the _real data_ storage to which they are the interface?

Think about the global `state` object from a front-end perspective (as in React or Vue). The "state" is the "object of all possibilities", and it starts as `{ }`. An empty object. Which has the potential to contain anything and everything.

We need to figure out a state schema/structure which represents the revisions and serves as a data mapper between them and the actual data location.

While the revisions (table definitions) themselves **and** their corresponding APIs are immutable and easy to rollback and go forward... the "data mapper" is mutable, only "goes forward" and represents the current state, reflecting the actual data storage situation.

The data mapper allows that an immutable API (revision 2, "getProjectName") with immutable code ("r2 get project name for ID 32") will retrieve the piece of data from where it currently is.

Because the "project name" piece of data may have started at a given table... but then moved to a foreign table. It is not where it was before, anymore.

When a new revision alters a previous one, it may transform a data column in a looked up data from other table (foreign key + field name). Our code will have to do two things:

- The actual database schema change
- The "data mapper" update

At the end, we have:
- multiple APIs/interfaces/views as immutable "table definition revisions" (command sequences)
- a JSON read/write data mapper which keeps old and new revisions pointing to same data source
- the real data source (database, containing the single source of data)

<hr>

While the sketch above sounds brilliant to me, and I feel enthusiastic as I finish to write it up... I am ready to acknowledge that I suffer some degree of [Dunning-Kruger effect](https://en.wikipedia.org/wiki/Dunning%E2%80%93Kruger_effect), and accept I missed important points, presenting an unfeasible "solution".

:-) At least, I had a good time on trying to face this challenge.
